bullet/src/JointFeatures.cc

```diff
@@ -27,51 +27,206 @@ namespace bullet {
 double JointFeatures::GetJointPosition(
     const Identity &_id, const std::size_t _dof) const
 {
-  (void) _id;
   (void) _dof;
-  double result = 0;
+  double result = ignition::math::NAN_D;
   if (this->joints.find(_id.id) != this->joints.end())
   {
-    btHingeAccumulatedAngleConstraint* hinge =
-      static_cast<btHingeAccumulatedAngleConstraint*>(this->joints.at(_id.id)->joint);
-    if (hinge)
+    const JointInfoPtr &jointInfo = this->joints.at(_id.id);
+    const int jointType = jointInfo->constraintType;
+    // Check the type of joint and act accordignly
+    if (jointInfo->constraintType ==
+        static_cast<int>(::sdf::JointType::REVOLUTE))
+    {
+      btHingeAccumulatedAngleConstraint* hinge =
+        static_cast<btHingeAccumulatedAngleConstraint*>(jointInfo->joint);
+      if (hinge)
+      {
+        result = hinge->getAccumulatedHingeAngle();
+        // result -= this->angleOffset;
+      }
+      else
+      {
+        ignerr << "Corrupted joint at index:" << _id.id << "\n";
+      }
+    }
+    else
     {
-      result = hinge->getAccumulatedHingeAngle();
+      ignerr << "Not a valid constrating type: " << jointType << "\n";
     }
-    // result -= this->angleOffset;
   }
-  // ignerr << "Position: " << _id.id << " -> " << result << std::endl;
+  igndbg << "Position: " << _id.id << " -> " << result << std::endl;
   return result;
 }
 
 /////////////////////////////////////////////////
 double JointFeatures::GetJointVelocity(
     const Identity &_id, const std::size_t _dof) const
 {
-  (void) _id;
   (void) _dof;
-  igndbg << "Dummy function GetJointVelocity\n";
-  return 0.0;
+  double result = ignition::math::NAN_D;
+  if (this->joints.find(_id.id) != this->joints.end())
+  {
+    const JointInfoPtr &jointInfo = this->joints.at(_id.id);
+    const int jointType = jointInfo->constraintType;
+    if (jointInfo->constraintType ==
+        static_cast<int>(::sdf::JointType::REVOLUTE))
+    {
+      btHingeAccumulatedAngleConstraint* hinge =
+        static_cast<btHingeAccumulatedAngleConstraint*>(jointInfo->joint);
+      if (hinge)
+      {
+        result = 0.0;
+        // Get the axis of the joint
+        btVector3 vec =
+          hinge->getRigidBodyA().getCenterOfMassTransform().getBasis() *
+          hinge->getFrameOffsetA().getBasis().getColumn(2);
+
+        math::Vector3 globalAxis(vec[0], vec[1], vec[2]);
+
+        if (this->links.find(jointInfo->childLinkId) != this->links.end())
```

Comment:
I'm not sure about these verifications, if jointInfo has an invalid childLink, we should explode ASAP, as probably something went really wrong

Reply:
These verifications check for joints with a parent/child link that doesn't have an extra child/parent link (attached to the world). The final angular velocity is the sum of the angular velocities of both links, the extra check is to see if it's attached to the world (although that's not really supported yet in constructionSdf)

Reply:
Checking if childLinkId is in this->links means checking if the link exist as an entity. Joints with the world as a parent/child would have the world identity as the id. I think that checking if the id is the same as the world id would be the proper check, but I'm also fine with the answer